toc/rfc/rfc-draft-bindable-secrets.md

```diff
@@ -0,0 +1,254 @@
+
+```console
+curl "https://api.example.org/v3/secrets/[guid]/value" \
+  -X GET \
+  -H "Authorization: bearer [token]"
+  
+"value" : {
+  "some-json-key" : "some-json-value"
+}
+```
+
+To support this, Cloud Controller will need to have both read and write permissions for the `secret`'s corresponding credential in CredHub.
+
+### Creating a `secret_binding`
+To create `secret_bindings`, users will use a new CLI command: `cf bind-secret`. This will create a `POST` request to `/v3/secret_bindings`.
+
+```console
+cf bind-secret APP_NAME SECRET_NAME BINDING_NAME
+```
+
+### CLI Changes
+Work will need to be done to implement the following commands:
+- `cf create-secret`
+- `cf bind-secret`
+- `cf secrets`
+- `cf secret`
+- `cf delete-secret`
+- `cf unbind-secret`
+
+In addition, modifications may be made to existing CLI commands to surface `secrets` and `secret_bindings`. For example, adding bound `secret`s to `cf app` output.
+
+### Cloud Controller Changes
+New database tables will need to be created to store `secret`s and `secret_binding`s. CRUD APIs will need to be created for these resources. We will also need an API to support viewing a `secret`'s value.
+
+### BBS changes
+BBS was recently updated to support `ServiceBindingFiles` as part of RFC 0030. A `secret`'s value may result in one or more files being created in the `tmpfs` mounted directory defined by the `MountPath` parameter.
+
+```
+action := &models.RunAction{
+  Path: "/path/to/executable",
+  Args: []string{"some", "args to", "pass in"},
+  Dir: "/path/to/working/directory",
+  User: "username",
+  EnvironmentVariables: []*models.EnvironmentVariable{
+    {
+      Name: "ENVNAME",
+      Value: "ENVVALUE",
+    },
+  },
+ ServiceBindingFiles: []*models.Files{
+    {
+      Name: "/etc/cf-instance-binding",
+      Value: "VALUE",
+    },
+  },
+  Secrets: []*models.Secret{
+    {
+      Name: "SECRETNAME",
+      MountPath: "/etc/conf/",
+      Value: "credhub-credential-id:<CREDHUB_CREDENTIAL_ID>",
+    },
+  },
+  ResourceLimits: &models.ResourceLimits{
+    Nofile: 1000,
+  },
+  LogSource: "some-log-source",
+  SuppressLogOutput: false,
+}
+```
+
+### Launcher Changes?
+The launcher (or something else) will need to be updated to retrieve credentials from CredHub for `Secret Bindings`. It currently already does this for CredHub references contained within `VCAP_SERVICES` using the app container's instance identity credentials. Is there something other than the launcher that should have this responsibility?
+
+Alternatively, credentials could be fetched by Cloud Controller (it does this already for Service Keys) and passed through directly to Diego. This has the disadvantage of increasing the size of the DesiredLRP and storing storing the secret contents in Diego's database.
+
+### Other Considerations
```

Comment:
Any thoughts about how secrets would be managed in app/space manifests?

Reply:
> Whether or not secrets themselves should be defined in a v2 space manifest is in another question.

@tcdowney what do you mean by this?

Reply:
@Gerg is working on a draft proposal of a "v2" version of the CF App manifest that is more declarative and includes the ability to configure other space scoped resources. This is still WIP, but he has a version of it here:
https://github.com/Gerg/community/blob/rfc_v2-app-manifests/toc/rfc/rfc-draft-v2-app-manifests.md#add-secrets